src/mpi/comm/commutil.c

```diff
@@ -514,6 +514,28 @@ static int check_hierarchy(MPIR_Comm * comm)
     comm->node_comm = NULL;
     comm->node_roots_comm = NULL;
 
+    if (num_nodes == 1) {
+        comm->hierarchy_flags |= MPIR_COMM_HIERARCHY__SINGLE_NODE;
+    } else if (num_local == 1) {
+        comm->hierarchy_flags |= MPIR_COMM_HIERARCHY__NO_LOCAL;
+    } else {
+        bool is_node_consecutive = true;
+        for (int i = 1; i < comm_size; i++) {
+            if (internode_table[i] < internode_table[i - 1]) {
+                is_node_consecutive = false;
+                break;
+            }
+        }
+
+        if (is_node_consecutive) {
+            comm->hierarchy_flags |= MPIR_COMM_HIERARCHY__NODE_CONSECUTIVE;
+        }
+
+        if (num_local * num_nodes == comm_size) {
+            comm->hierarchy_flags |= MPIR_COMM_HIERARCHY__NODE_BALANCED;
+        }
```

Comment:
is this check strict enough? the old check put all the ranks in node buckets and checked that each bucket was the same size.